lib/switchtec.c

```diff
@@ -880,26 +950,45 @@ static int write_parsed_log(struct log_a_data log_data[],
 		hours = time % 24;
 		days = time / 24;
 
-		/*
-		 * log entry number, module ID, and log severity are in the
-		 * 3rd DWord
-		 */
-		entry_num = log_data[i].data[2] & 0x0000FFFF;
-		mod_id = (log_data[i].data[2] >> 16) & 0xFFF;
-		log_sev = (log_data[i].data[2] >> 28) & 0xF;
-
-		/* validate the module ID */
-		if ((mod_id > defs->num_alloc) ||
-		    (strlen(defs->module_defs[mod_id].mod_name) == 0)) {
-			if (fprintf(log_file, "(Invalid module ID: 0x%x)\n",
-				    mod_id) < 0)
-				goto ret_print_error;
-			continue;
+		if (log_type == SWITCHTEC_LOG_PARSE_TYPE_APP) {
+			/*
+			 * app log: module ID and log severity are in the 3rd
+			 * DWord
+			 */
+			mod_id = (log_data[i].data[2] >> 16) & 0xFFF;
+			log_sev = (log_data[i].data[2] >> 28) & 0xF;		
+
+			if ((mod_id > defs->num_alloc) ||
+			    (defs->module_defs[mod_id].mod_name == NULL) ||
```

Comment:
I still don't understand why this line is being snuck into this patch...

Reply:
Ah, right. I was trying out passing mailbox files but specifying the type as app log, which was leading to a segfault. I added this line as a test, which fixed it, but it's a bit of a hack and should probably be removed.

I'm not sure if we want to handle to case of the user specifying the wrong log type (app / mailbox) for the input files. If we do it would probably require some more advanced parsing of the log definition file.

Reply:
Ah, well you can add that line but do it in a different patch with justification.

But clearly there is a bigger problem, if code can get this far when you pass it an invalid data file and that needs to be addressed properly.

Reply:
Sure, I'll move this addition to a new patch.

I agree that it would be better to validate the data earlier. That's why I originally preferred parsing the definition file using something like regular expressions or sscanf, as we can at least check for the proper data types (e.g. the module description is a string, a 4-digit hex number, and a decimal number). We can deal with this in a later PR.

Reply:
I don't see why sscanf or re would be any different. You can get the same error checking, it just wasn't done.

Seems like we are missing some error checking on the strtol and it might make sense to error out if strtok doesn't find the requisite number of fields instead of ignoring the line.

The mailbox file looks a bit worse as it seems to just store every line regardless of what's in the file. Can we maybe at least at a header to that file so this code can at least verify something?

Reply:
I'll update the strtol error checking, which should catch the unexpected values. Lines are ignored if the strtok fails as some lines are blank. I can instead explicitly skip blank lines and error out if strtok fails.

The mailbox definition file is just a raw list of error strings, which makes it difficult to validate. We could potentially add a header, but I don't think we'd want to rely on that in case users pass in earlier versions that don't have the header.

Reply:
Maybe add a force option to accept old mailbox files that don't have a header?

Reply:
I've updated the strol() error checking. The mod_name NULL check is no longer needed.

We can add headers to the definition files, but this will need to be approved and implemented on the firmware side first. If we're going ahead with that then I can implement the header checking in a separate PR.